pallets/cash/src/internal/liquidate.rs

```diff
@@ -0,0 +1,553 @@
+// Note: The substrate build requires these be re-exported.
+pub use our_std::{
+    cmp::{max, min},
+    collections::btree_set::BTreeSet,
+    convert::{TryFrom, TryInto},
+    fmt, if_std, result,
+    result::Result,
+    str,
+};
+
+use frame_support::storage::{StorageMap, StorageValue};
+
+use crate::{
+    chains::ChainAccount,
+    core::{self, get_price, get_value},
+    effectful::Effectful,
+    factor::Factor,
+    internal::liquidity,
+    params::MIN_TX_VALUE,
+    reason::Reason,
+    require, require_min_tx_value,
+    types::{AssetInfo, AssetQuantity, CashPrincipalAmount, CASH},
+    Config, Event, GlobalCashIndex, Module, TotalBorrowAssets, TotalSupplyAssets,
+};
+
+pub fn liquidate_internal<T: Config>(
+    asset: AssetInfo,
+    collateral_asset: AssetInfo,
+    liquidator: ChainAccount,
+    borrower: ChainAccount,
+    amount: AssetQuantity,
+) -> Result<(), Reason> {
+    require!(asset != collateral_asset, Reason::InKindLiquidation); // TODO: Can we allow this now?
+    require_min_tx_value!(get_value::<T>(amount)?);
+
+    let liquidation_incentive = Factor::from_nominal("1.08"); // XXX spec first
+    let seize_amount = amount
+        .mul_factor(liquidation_incentive)?
+        .mul_price(get_price::<T>(asset.units())?)?
+        .div_price(
+            get_price::<T>(collateral_asset.units())?,
+            collateral_asset.units(),
+        )?;
+
+    require!(
+        !liquidity::has_non_negative_liquidity::<T>(borrower)?,
```

Comment:
Probably worth just adding a `has_negative_liquidity` wrapper? I think require double negative is harder to parse

Reply:
I might rework these once we get lock and extract moved over, since they can mostly all go away at that point.